src/lib/statistics/chunk_statistics/histograms/abstract_histogram.cpp

```diff
@@ -0,0 +1,620 @@
+      }
+
+      // If the pattern starts with a MatchAll, we can not prune it.
+      if (value.front() == '%') {
+        return false;
+      }
+
+      /**
+       * We can prune prefix searches iff the domain of values captured by a prefix pattern is prunable.
+       *
+       * Example:
+       * bins: [a, b], [d, e]
+       * predicate: col LIKE 'c%'
+       *
+       * With the same argument we can also prune predicates in the form of 'c%foo',
+       * where foo can be any pattern itself.
+       * We only have to consider the pattern up to the first AnyChars wildcard.
+       */
+      const auto match_all_index = value.find('%');
+      if (match_all_index != std::string::npos) {
+        const auto search_prefix = value.substr(0, match_all_index);
+        if (can_prune(PredicateCondition::GreaterThanEquals, search_prefix)) {
+          return true;
+        }
+
+        const auto search_prefix_next_value = next_value(search_prefix, _supported_characters, search_prefix.length());
+
+        // If the next value is the same as the prefix, it means that there is no larger value in the domain
+        // of substrings. In that case we cannot prune, because otherwise we previous check would already return true.
+        if (search_prefix == search_prefix_next_value) {
+          return false;
+        }
+
+        if (can_prune(PredicateCondition::LessThan, search_prefix_next_value)) {
+          return true;
+        }
+
+        const auto search_prefix_bin = _bin_for_value(search_prefix);
+        const auto search_prefix_next_value_bin = _bin_for_value(search_prefix_next_value);
+
+        if (search_prefix_bin == INVALID_BIN_ID) {
+          const auto search_prefix_upper_bound_bin = _upper_bound_for_value(search_prefix);
+
+          // In an EqualElementCountHistogram, if both values fall into the same gap, we can prune the predicate.
+          // We need to have at least two bins to rule out pruning if search_prefix < min
+          // and search_prefix_next_value > max.
+          if (search_prefix_next_value_bin == INVALID_BIN_ID && bin_count() > 1ul &&
+              search_prefix_upper_bound_bin == _upper_bound_for_value(search_prefix_next_value)) {
+            return true;
+          }
+
+          // In an EqualElementCountHistogram, if the search_prefix_next_value is exactly the lower bin edge of
+          // the upper bound of search_prefix, we can also prune.
+          // That's because search_prefix_next_value does not belong to the range covered by the pattern,
+          // but is the next value after it.
+          if (search_prefix_next_value_bin != INVALID_BIN_ID &&
+              search_prefix_upper_bound_bin == search_prefix_next_value_bin &&
+              _bin_min(search_prefix_next_value_bin) == search_prefix_next_value) {
+            return true;
+          }
+        }
+
+        // In an EqualWidthHistogram, if both values fall into a bin that has no elements,
```

Comment:
See above. I have a bad feeling about all these "In a(n) ...Histogram" sentences in the *abstract* histogram.

Reply:
See above.